### ISSUE TITLE
Pure float32: 20 derived properties are non-finite though their float64 values are representable — four consumer families read a raw erg/s key whose finite log companion already exists

### ISSUE THREAD

**Issue by suchethac** [suchethac/tengri#1837]
**Child of #1206** — the concrete, measured instance of the still-unchecked §1 bullet *"Energy-balance / bolometric integrals — `∫ L_ν dν` (~1e47 erg/s) overflow in pure float32"*, plus two sites of the §2 *"published scalars inherently float32-unrepresentable"* class.

## Summary

Under `jax.config.update("jax_enable_x64", False)`, **20 derived properties are `inf` or `nan` even though their float64 values are perfectly representable in float32** (they are magnitudes, dex ratios, Gyr, and L☉ — all between 1e-1 and 1e10).

Every one of the 20 is a **consumer defect, not a missing capability**: the range-safe route it needs *already exists* in the same forward pass. This is a partial migration — the #1206 remedy was applied to some call sites and not to their neighbours.

The clearest single instance is inside one function, `state_to_sed_quantities`:

```python
# forward/component_factory.py:793-794
# ... folds 1/L_sun into the integral, so the ~1e43 erg/s
# value is never formed (float32-safe, #1206).
l_bol = jnp.abs(compute_bolometric_luminosity(sed, wave))   # ← remedy applied

# forward/component_factory.py:799, five lines later
l_absorbed = jnp.asarray(derived.get("L_absorbed", 0.0))    # ← remedy not applied
l_dust_absorbed = l_absorbed / L_SUN
```

In the same float32 pass, the **larger** quantity survives and the **smaller** one dies:

| quantity | float64 [L☉] | float32 | route |
|---|---|---|---|
| `l_bol` | 1.124e+10 | **1.123770e+10** finite | `_trapz_to_lsun` (peak-factored) |
| `l_dust_absorbed` | 9.454e+09 | **`inf`** | raw `L_absorbed` [erg/s] |

Magnitude cannot explain that ordering — float32 loss degrades monotonically with size. Big value fine, smaller value `inf` can only mean the two took different code paths.

## Reproduction

Pure float32 must be selected **after** `import tengri` — `src/tengri/__init__.py:100` runs an unconditional `jax.config.update("jax_enable_x64", True)`, which silently overrides `JAX_ENABLE_X64=0` from the environment. (Filed separately.)

```python
import jax
from tengri import FIXED, Fixed, Observation, Photometry, SEDModel, Uniform
from tengri.components.stellar.sps.dsps_wrapper import load_ssp_data

jax.config.update("jax_enable_x64", False)   # <- after the import, not via env

ssp = load_ssp_data("data/fsps_prsc_miles_chabrier.h5")
model = SEDModel.build(
    ssp_data=ssp,
    observation=Observation(photometry=Photometry.from_names(
        ["sdss_u", "sdss_g", "sdss_r", "sdss_i", "sdss_z"])),
    sfh={"type": "delayed", "all_params": FIXED,
         "log_total_mass": Uniform(9.0, 11.0), "tau_gyr": 1.5, "age_gyr": 5.0},
    dust={"type": "two_component", "law_bc": "calzetti", "all_params": FIXED,
          "tau_diff": Uniform(0.0, 1.5), "tau_bc": 0.3},
    neb={"type": "cue", "all_params": FIXED},
    redshift=Fixed(0.1), approx=None,
)
pred = model.predict({"sfh_delayed_log_total_mass": 10.0, "dust_tau_diff": 0.5})
pred.properties["l_dust_absorbed"]   # inf   (float64: 1.134e+10 L_sun)
pred.properties["irx"]               # nan   (float64: -0.6745 dex)
pred.properties["m_uv"]              # inf   (float64: -16.87 mag)
```

Measured on `23bf05280` with SSP `fsps_prsc_miles_chabrier.h5`, z = 0.1, `approx=None`, against a float64 arm with identical pinned parameters and identical pinned mock data. The float32 arm was gated non-vacuous first (`jax_enable_x64=False`, SED dtype `float32`, SSP dtype `float32`) — without that gate the arm silently runs in float64.

## The 20 failures

Two model chains: `stellar+dust` (7) and `stellar+dust+Cue` (13).

```
irx                              f64 = -8.434e-01 / -6.745e-01     both chains
irx_fuv                          f64 = -7.437e-01 / -6.460e-01     both chains
l_dust_absorbed                  f64 =  9.454e+09 /  1.134e+10     both chains
luminosity_weighted_age_gyr      f64 =  1.064e+00                  both chains
luminosity_weighted_metallicity  f64 =  0.000e+00                  both chains
m_uv                             f64 = -1.683e+01 / -1.687e+01     both chains
rest_uv_color                    f64 =  1.393e+00 /  1.288e+00     both chains
balmer_decrement                 f64 =  2.789e+00                  Cue only
bpt_nii                          f64 = -5.057e-01                  Cue only
bpt_sii                          f64 = -4.628e-01                  Cue only
o32                              f64 = -4.503e-01                  Cue only
o3hb                             f64 =  7.665e-02                  Cue only
r23                              f64 =  6.954e-01                  Cue only
```

**Explicitly excluded** from this issue: the 11 raw emission-line luminosities and the X-ray luminosities, whose float64 values (~1e41–1e43 erg/s) genuinely *exceed* float32's 3.4028235e38 ceiling. Those need the breaking unit change already tracked as #1206 §3. The 20 above are the strictly harder case — **the answer is representable and only the path is not.**

### A. Energy balance — `l_dust_absorbed` (2 instances)

Reads `state.derived["L_absorbed"]`, a `DerivedKey` declared in **erg/s** (`forward/orchestrator.py:233`). Measured in float32: `L_absorbed` **0.0% finite**, while its companion `log_L_ir` is **100% finite** at 43.5586 dex — and 10^(43.5586 − log₁₀L☉) = 9.46e9 L☉, reproducing the float64 answer.

- `forward/component_factory.py:799`
- `components/stellar/component.py:3553` (`_l_dust_absorbed_fn`)

The companion is already the preferred read at three *other* sites — `components/agn/component.py:389`, `components/dust/emission/_component_base.py:95`, `components/radio/component.py:287` — which is what makes this a partial migration rather than a missing feature.

### B. UV diagnostics — `irx`, `irx_fuv`, `m_uv`, `rest_uv_color` (8 instances)

Two independent sub-mechanisms, both measured:

**B1 — erg/s products.** Each overflows on its own, so fixing either alone leaves the property broken:

| expression | site | float64 | float32 |
|---|---|---|---|
| `l_tir_lsun * L_SUN` | `utils/sed_quantities.py:561` (`compute_irx`) | 7.213740e+41 | **`inf`** |
| `nu_1600 * l_nu_1600` | `utils/sed_quantities.py:543` (`compute_uv_luminosity_1600`) | 5.030042e+42 | **`inf`** |

`compute_irx` then evaluates `inf / inf` → **`nan`**. Note `compute_l_tir` itself is *finite* (1.884457e+08 L☉) because it goes through `_trapz_to_lsun` — `compute_irx` immediately multiplies that safe value back into erg/s.

**B2 — magnitude zero-points.** This one is worth reading closely: the **inputs and outputs are both comfortably representable, and only an internal constant is not.**

```
lnu_1500_band   f64 = 2.352914e+27   f32 = 2.352915e+27   ok
m_uv            f64 = -1.683417e+01  f32 = inf            BREAKS HERE
f_u_band        f64 = 8.007146e+27   f32 = 8.007147e+27   ok
mag_u           f64 = -1.183586e+02  f32 = -inf           BREAKS HERE
rest_uv_color   f64 = 1.393216e+00   f32 = nan            (-inf) - (-inf)
```

The constants, verified numerically against `float32 max = 3.4028235e38`:

- `utils/magnitudes.py:160` — `4.0 * jnp.pi * TEN_PC_CM**2`. `TEN_PC_CM**2 = 9.521406e38` already overflows **at the square**, before the 4π; the full factor is 1.196495e40.
- `utils/magnitudes.py:89` — `fnu_safe / MAGGIES_ZP_CGS` with `MAGGIES_ZP_CGS = 3.631e-20`. For the measured band means this is 2.205218e+47 … 7.956792e+47, all over the ceiling.

There is a **second-order defect stacked on B2**: after `lnu / inf → 0.0`, the guard at `utils/magnitudes.py:88` is `jnp.maximum(fnu_cgs, 1e-300)`, and `1e-300` is below float32's smallest subnormal (1.401e-45), so it is an exact no-op there — `log10(0.0) = -inf` → `-2.5 × -inf = +inf`, giving the measured `m_uv = inf`. This is the class swept in #1492 (closed), at a site that sweep did not reach.

### C. Line-ratio diagnostics — `balmer_decrement`, `bpt_nii`, `bpt_sii`, `o32`, `o3hb`, `r23` (6 instances)

All read `state.derived["line_lums"]` (erg/s), measured **17.2% finite** in float32, while `log_line_lums` is **100% finite** in the same pass. Seven raw read sites on current main:

- `components/nebular/component.py:1045` — `_line_luminosity_helper`, the shared choke point
- `components/nebular/component.py:1115, 1134, 1155, 1174, 1196, 1215` — one direct read per ratio property

These ratios are scale-invariant, so they lose nothing by being computed in the log domain.

### D. Luminosity-weighted quantities — `luminosity_weighted_age_gyr`, `luminosity_weighted_metallicity` (4 instances)

`forward/component_factory.py:845` reads raw `L_age` (erg/s per SSP age bin), measured **85 of 93 bins `inf`** in float32 — float64 peak is 3.345881e+42 against a 3.4028235e+38 ceiling. The subsequent `jnp.sum(ssp_ages_yr * L_age)` multiplies by ~1e10 on top, so it overflows twice over.

`log_L_age` is present and **correct**, which I checked explicitly because a naive `~isfinite` count makes it look broken (it reports 91.4% finite):

- all 8 non-finite entries are **exactly `-inf`**, none `NaN`
- float64 `L_age` is **exactly 0.0** at all 8 → `log10(0) = -inf` is the right answer
- the non-finite masks are **identical** between the float32 and float64 arms
- the finite 85 agree with float64 to **1.522e-07**

These are ratios of sums, hence scale-invariant — normalising `L_age` by its peak leaves the answer unchanged.

## Root cause in one line

`log_L_ir`, `log_L_age`, `log_line_lums` and `log_nion` are all declared contract keys and all **100% finite** in float32. Some consumers were migrated to them; these four families were not. Only `L_age` needed a second look, and it too turned out to be a consumer defect.

## Suggested remedy

The pattern already exists in `utils/sed_quantities.py:197` (`_trapz_to_lsun`, written for #1206): never materialise the erg/s value — factor it as `norm * pow10(log10(peak) - LOG10_L_SUN)`.

1. **A, C, D** — read the `log_*` companion where present and reconstruct via `pow10(log_x - offset)`, or (for C and D, which are scale-invariant) normalise by the peak and skip the reconstruction entirely.
2. **B1** — keep `compute_irx` in L☉/dex; do not multiply a range-safe L☉ value back into erg/s. Give `compute_uv_luminosity_1600` a log-domain form, or an L☉ one matching `compute_l_tir`.
3. **B2** — factor the distance modulus and the AB zero-point in the log domain in `utils/magnitudes.py` (both are pure constants), and replace the `1e-300` floor with `representable_floor(...)`, which `components/nebular/component.py:1030` already uses for exactly this reason.

Since these consumers sit next to already-migrated ones, a **guard that enumerates every raw read of a declared erg/s `DerivedKey`** would be worth more than fixing the individual sites — that is what would have caught this set, and what would catch the next one.

## Scope — what is **not** broken

Measured on the same two chains, same arms, so the fix stays targeted:

| stage | result | verdict |
|---|---|---|
| Forward photometry | max rel err **2.631e-06** / **3.934e-06** | correct |
| χ² gradients | all finite, rel err **1.29e-05 – 1.34e-04** | correct |
| MAP fits | converge; losses agree to 6 digits (3.357571 / 3.357558 and 1.925553 / 1.925540), params to ~1e-5, in 3.0 s / 5.0 s | correct |

**The fitting path is sound in pure float32.** This issue is confined to derived-property read-out.

**Comment by suchethac** [comment]
**Correction to the "Scope — what is not broken" section of this issue.**

That section ends with "**The fitting path is sound in pure float32.**" That claim is too strong and should not have been written from the evidence behind it. Retracting it here rather than editing it silently.

The gradient check it rests on freed exactly two parameters:

```python
FREE = ["sfh_delayed_log_total_mass", "dust_tau_diff"]
```

Both are well-conditioned scale parameters with gradients of order 1e2, on a delayed-tau SFH, with redshift `Fixed`. A probe like that cannot speak to shape parameters, to a richer free set, or to photo-z — and I generalized it into a statement about fitting in general.

## What is actually established

Re-measured on `a0ea35d16`, DPL SFH, 6 free parameters, GALEX+SDSS+WISE, `approx=None`, A/A control 0.000e+00 in both arms:

| parameter | f64 | f32 | rel err |
|---|---|---|---|
| `sfh_dpl_log_total_mass` | -1.721296e+02 | -1.722221e+02 | 5.4e-04 |
| `sfh_dpl_alpha` | 5.875396e+01 | 5.879205e+01 | 6.5e-04 |
| `sfh_dpl_beta` | -5.011727e+00 | -5.014171e+00 | 4.9e-04 |
| `sfh_dpl_tau_gyr` | -2.595376e+01 | -2.596825e+01 | 5.6e-04 |
| `dust_tau_diff` | 4.854857e+01 | 4.861930e+01 | 1.5e-03 |
| `met_logzsol` | 1.773480e+01 | 1.776003e+01 | 1.4e-03 |

Adding a free redshift to that same model, everything else identical:

```
[f64/freez]  redshift  g =  3.121449e+02
[f32/freez]  redshift  g =  nan
```

The six components above are unchanged and still accurate; only `redshift` is NaN.

## Two distinct failure modes, and they need different fixes

1. **The redshift gradient is NaN in float32.** Reproducible, and isolated — the other six components in the same pass are fine. This is not a conditioning problem: the float64 value here is 3.12e+02, a large healthy gradient, and catastrophic cancellation produces a wrong *finite* number, never a NaN. It looks like #1415 (`apply_log10_scale` underflowing in reverse at the flux-projection seam), which is the parameter this gradient flows through.

2. **Ill-conditioned points lose float32 gradients to cancellation.** Separately reported at an operating point where the float64 `sfh_dpl_tau_gyr` gradient is -3.428e-02 beside neighbours of 40-300; there the float32 value came back -2.897e+00. That is a genuine near-total cancellation and float32 cannot carry it. Note the contrast with the table above, where the same parameter has a float64 gradient of -2.60e+01 — about 750x larger — and float32 tracks it to 5.6e-04. The failure is a property of the operating point, not of the parameter.

## Corrected scope

- Forward photometry in float32: **sound** (max rel err 2.6e-06 / 3.9e-06).
- Derived properties: **this issue**, fixed by #1847.
- Fixed-redshift gradients: **accurate at well-conditioned points**, and unreliable wherever a component's gradient sits orders of magnitude below its neighbours. Not a blanket pass or fail.
- **Photo-z (free redshift) in float32: broken.** The redshift gradient is NaN, so no gradient-based fit with a free redshift can work.
- MAP fits: the earlier "converge, losses agree to 6 digits" result stands only for the 2-parameter delayed-tau model it was measured on.

Since JAX-Metal was wanted for *fitting*, item 4 is the one that matters for #1206's acceptance criterion, and it is not fixed by #1847. #1847 is confined to derived-property read-out, as its own scope section says; nothing above changes what that PR does.
